00_Base/src/interfaces/repository.ts

```diff
@@ -3,21 +3,66 @@
 //
 // SPDX-License-Identifier: Apache 2.0
 
+import { EventEmitter } from 'events';
+
+export type CrudEvent<T> = {
+  created: [T[]];
+  updated: [T[]];
+  deleted: [T[]];
+};
+
 /**
  * Represents a generic CRUD repository.
  *
  * @template T - The type of the values stored in the repository.
  */
-export interface ICrudRepository<T> {
+export abstract class CrudRepository<T> extends EventEmitter {
```

Comment:
what made you choose to extend EventEmitter instead of just having an event emitter on the repository class as a property? My thinking is you can just have something like 
```
export abstract class CrudRepository<T> extends EventEmitter {
   ...
   eventEmitter = new EventEmitter();
   ...
   public async create(value: T, namespace?: string): Promise<T> {
    const result = await this._create(value, namespace);
    this.eventEmitter.emit('created', [result]);
    return result;
  }
}
```
and I think then you dont need these:
```
on<K extends keyof CrudEvent<T>>(
    event: K,
    listener: (...args: CrudEvent<T>[K]) => void,
  ): this {
    return super.on(event, listener as (...args: any[]) => void);
  }
  
  emit<K extends keyof CrudEvent<T>>(
    event: K,
    ...args: CrudEvent<T>[K]
  ): boolean {
    return super.emit(event, ...args);
  }
  ```

Reply:
That would no longer supply typing for the emitted events, or require the use of CrudEvent keys. This would make it harder to develop with and less safe simultaneously. [See this comment for more detail.](https://github.com/citrineos/citrineos-core/pull/87#discussion_r1602183957)